Client/Frontend/Brave Today/Composer/FeedDataSource.swift

```
@@ -0,0 +1,544 @@
+    /// example: `sources.json`)
+    ///
+    /// Cache lifetime will be based on the modification date of the cached file. Data downloaded from the web
+    /// will only be cached if it is successfully decoded into the given `DataType`.
+    private func loadResource<DataType>(
+        _ resource: TodayResource,
+        decodedTo: DataType.Type
+    ) -> Deferred<Result<DataType, Error>> where DataType: Decodable {
+        let filename = resource.filename
+        let fileManager = FileManager.default
+        let deferred = Deferred<Result<Data, Error>>(value: nil, defaultQueue: .main)
+        let cachedPath = fileManager.getOrCreateFolder(name: Self.cacheFolderName)?.appendingPathComponent(filename).path
+        if let cachedPath = cachedPath,
+            !isResourceExpired(resource) {
+            todayQueue.async {
+                if let cachedContents = fileManager.contents(atPath: cachedPath) {
+                    deferred.fill(.success(cachedContents))
+                }
+            }
+        } else {
+            session.dataRequest(with: Self.baseURL.appendingPathComponent(filename)) { data, response, error in
+                if let error = error {
+                    deferred.fill(.failure(error))
+                    return
+                }
+                guard let data = data else { return }
+                deferred.fill(.success(data))
+            }
+        }
+        return deferred.mapQueue(todayQueue) { result in
+            switch result {
+            case .success(let data):
+                do {
+                    let decodedSources = try self.decoder.decode(DataType.self, from: data)
+                    if !data.isEmpty {
+                        if !FileManager.default.writeToDiskInFolder(data, fileName: filename, folderName: Self.cacheFolderName) {
+                            logger.error("Failed to write sources to disk")
```

we should return `.failure` here too I think

---

Im not sure if a failure to write to cache warrants entire today feed to show a failure? If we were to do that we should surface the proper error to the user (such as out of space or something)